# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [95]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [96]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [97]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [98]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [99]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [100]:
poke = pd.read_csv("pokemon.csv")

In [101]:
poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [102]:
poke["Type 1"].value_counts(normalize = False)


Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [103]:
poke["Type 1"].value_counts(normalize = True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [104]:
pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

pd.crosstab(poke["Stage"], poke["Legendary"])

Legendary  False  True
Stage                 
1             75     4
2             56     0
3             16     0

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [105]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [106]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [107]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [108]:
poke["Attack"].mean()

# the average value of Attack is 72.54

72.54966887417218

In [109]:
poke["Defense"].median()


65.0

In [110]:
poke["Speed"].mode()

0    70
1    90
dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [111]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [112]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [113]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [114]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [115]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [116]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [117]:
poke.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Num        151 non-null    int64 
 1   Name       151 non-null    object
 2   Type 1     151 non-null    object
 3   Type 2     67 non-null     object
 4   Total      151 non-null    int64 
 5   HP         151 non-null    int64 
 6   Attack     151 non-null    int64 
 7   Defense    151 non-null    int64 
 8   SpAtk      151 non-null    int64 
 9   SpDef      151 non-null    int64 
 10  Speed      151 non-null    int64 
 11  Stage      151 non-null    int64 
 12  Legendary  151 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 14.4+ KB


In [118]:
poke["HP"].max() - poke["HP"].min()

240

In [119]:
poke["SpAtk"].mean()

67.13907284768212

In [120]:
poke["SpAtk"].std()

28.53419930191353

In [121]:
#For the Attack column, determine which value falls at the 80th percentile?
#Which value falls at the 15th percentile?
#For each of these values, write a single sentence describing what this means. 

np.percentile(poke["Attack"], 80)



95.0

In [122]:
np.percentile(poke["Attack"], 15)


45.0

Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [123]:
poke["Attack"].median()

70.0

In [124]:
poke["Attack"].quantile(.50)

70.0

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [125]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [126]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [127]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [128]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [129]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [130]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [131]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [132]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [133]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [134]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [135]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [136]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [137]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [138]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [139]:
ba = pd.read_excel("babies.xlsx")

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [140]:
ba.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

In [141]:
ba.shape

(1236, 7)

In [142]:
ba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


In [143]:
ba.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [144]:
ba.dropna(inplace = True)

In [145]:
# reset the index after removing the rows everytime
ba = ba.reset_index(drop = True)

In [146]:
ba.shape

(1174, 7)

In [147]:
Rows_Dropped = 1236 - 1174

print("Rows Dropped: ", Rows_Dropped)

Rows Dropped:  62


4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [ ]:
# parity and smoke are the two qualitative variable in the data set.
#However, parity is quantified by assigning 1 = Yes, and 0 = No for previous pregnancy
#smoke is quantified by assigning 1 = Yes for Smoker, 0 = No for non smokeer

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [148]:
ba["parity"].replace([0, 1], ["No", "Yes"], inplace = True)
ba.head()

bwt  gestation parity   age  height  weight  smoke
0  120      284.0     No  27.0    62.0   100.0    0.0
1  113      282.0     No  33.0    64.0   135.0    0.0
2  128      279.0     No  28.0    64.0    77.0    1.0
3  108      282.0     No  23.0    67.0   125.0    1.0
4  136      286.0     No  25.0    62.0    93.0    0.0

In [149]:
ba["smoke"].replace([0, 1], ["Non-smoker", "Smoker"], inplace = True)
ba.head()

bwt  gestation parity   age  height  weight       smoke
0  120      284.0     No  27.0    62.0   100.0  Non-smoker
1  113      282.0     No  33.0    64.0   135.0  Non-smoker
2  128      279.0     No  28.0    64.0    77.0      Smoker
3  108      282.0     No  23.0    67.0   125.0      Smoker
4  136      286.0     No  25.0    62.0    93.0  Non-smoker

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [150]:
baq = ba.copy()

In [151]:
baq.shape

(1174, 7)

In [152]:
q1 = baq["bwt"].quantile(.25)
q3 = baq["bwt"].quantile(.75)

print("Q1: ", q1)
print("Q3: ", q3)

Q1:  108.0
Q3:  131.0


In [153]:
IQR = q3 - q1
print("IQR: ", IQR)

IQR:  23.0


In [154]:
top = q3 + (IQR * 1.5)
bottom = q1 - (IQR * 1.5)

print("Upper Limit: ", top)
print("Lower Limit: ", bottom)

Upper Limit:  165.5
Lower Limit:  73.5


In [155]:

outliers = baq.loc[(baq["bwt"] > top) | (baq["bwt"] < bottom)].index

print("Index Values: ", outliers)

Index Values:  Int64Index([ 220,  287,  335,  430,  465,  492,  519,  557,  594,  663,  692,
             700,  780,  851,  859,  922,  964,  978, 1005, 1007, 1041, 1081,
            1090, 1111],
           dtype='int64')


In [156]:
baq = baq.drop(outliers)
baq.shape

(1150, 7)

In [157]:
# removing outliers from gestation variable

g_q1 = baq["gestation"].quantile(.25)
g_q3 = baq["gestation"].quantile(.75)

g_iqr = g_q3 - g_q1

g_top = g_q3 + (g_iqr * 1.5)
g_bottom = g_q1 - (g_iqr * 1.5)

print("Upper limit: ", g_top)
print("Lower limit: ", g_bottom)

Upper limit:  312.0
Lower limit:  248.0


In [158]:
g_outliers = baq.loc [(baq["gestation"] > g_top) | (baq["gestation"] < g_bottom)].index

print("Index Value: ", g_outliers)


Index Value:  Int64Index([   5,    6,    9,   56,   60,   63,  109,  119,  143,  173,  177,
             182,  193,  198,  200,  215,  221,  233,  238,  257,  321,  346,
             367,  409,  428,  449,  475,  487,  592,  641,  654,  664,  680,
             699,  713,  720,  729,  735,  779,  781,  784,  819,  873,  898,
             914,  946,  963,  969, 1016, 1076, 1082, 1084, 1088, 1094, 1114,
            1119, 1138, 1145, 1146, 1155, 1157, 1164],
           dtype='int64')


In [159]:
baq.iloc[[5,    6,    9,   56,   60,   ]]

bwt  gestation parity   age  height  weight       smoke
5   138      244.0     No  33.0    62.0   178.0  Non-smoker
6   132      245.0     No  23.0    65.0   140.0  Non-smoker
9   140      351.0     No  27.0    68.0   120.0  Non-smoker
56   75      232.0     No  33.0    61.0   110.0  Non-smoker
60  101      245.0     No  23.0    63.0   130.0      Smoker

In [160]:
baq = baq.drop(g_outliers)
print(baq.shape)

(1088, 7)


In [161]:
baq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1088 entries, 0 to 1173
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1088 non-null   int64  
 1   gestation  1088 non-null   float64
 2   parity     1088 non-null   object 
 3   age        1088 non-null   float64
 4   height     1088 non-null   float64
 5   weight     1088 non-null   float64
 6   smoke      1088 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 68.0+ KB


In [ ]:
# removing outliers for age variable

In [162]:
qa1 = baq["age"].quantile(.25)
qa3 = baq["age"].quantile(.25)

iqr_age = qa3 = qa1

top_age = qa3 + (iqr_age *1.5)
bottom_age = qa1 - (iqr_age *1.5)

print("Upper Limit: ", top_age)
print("Lower Limit: ", bottom_age)

Upper Limit:  57.5
Lower Limit:  -11.5


In [163]:
age_outliers = baq.loc[(baq["age"] > top_age) | (baq["age"] < bottom_age)].index

print("index: ", age_outliers)

index:  Int64Index([], dtype='int64')


In [164]:
baq = baq.drop(age_outliers)
baq.shape

(1088, 7)

In [ ]:
# removing outliers for height variable

In [165]:
qh1 = baq["height"].quantile(.25)
qh3 = baq["height"].quantile(.75)

iqr_h = qh3 - qh1

top_h = qh3 + (iqr_h * 1.5)
bottom_h = qh1 - (iqr_h * 1.5)

print("Upper Limit: ", top_h)
print("Lower Limit: ", bottom_h)

Upper Limit:  72.0
Lower Limit:  56.0


In [166]:
h_outliers = baq.loc[(baq["height"] > top_h) | (baq["height"] < bottom_h)].index
print("Index: ", h_outliers)

Index:  Int64Index([404, 1147], dtype='int64')


In [167]:
baq = baq.drop(h_outliers)

In [168]:
baq.shape

(1086, 7)

In [169]:
# reset the index after removing the rows everytime
baq = baq.reset_index(drop = True)

In [170]:
baq["height"].value_counts()

65.0    165
64.0    162
63.0    149
66.0    138
62.0    120
67.0     93
61.0     90
60.0     53
68.0     47
59.0     22
69.0     18
70.0     13
58.0      8
71.0      5
56.0      1
72.0      1
57.0      1
Name: height, dtype: int64

In [ ]:
# removing outliers for weight variable

In [171]:
qw1 = baq["weight"].quantile(.25)
qw3 = baq["weight"].quantile(.75)

iqr_w = qw3 - qw1

top_w = qw3 + (iqr_w * 1.5)
bottom_w = qw1 - (iqr_w * 1.5)

print("Upper Limit: ", top_w)
print("Lower Limit: ", bottom_w)

Upper Limit:  171.5
Lower Limit:  79.5


In [172]:
w_outliers = baq.loc[(baq["weight"] > top_w) | (baq["weight"] < bottom_w)].index

In [62]:
print("index: ", w_outliers)

index:  Int64Index([   2,   10,   17,   19,   23,   36,   78,  101,  108,  129,  131,
             140,  157,  158,  160,  189,  244,  266,  269,  298,  357,  372,
             398,  441,  443,  453,  457,  490,  521,  533,  547,  611,  612,
             634,  643,  744,  753,  754,  760,  779,  801,  812,  822,  889,
            1018, 1031, 1066, 1083],
           dtype='int64')


In [173]:
baq = baq.drop(w_outliers)
baq.shape

(1038, 7)

In [64]:
baq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1038 entries, 0 to 1085
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1038 non-null   int64  
 1   gestation  1038 non-null   float64
 2   parity     1038 non-null   object 
 3   age        1038 non-null   float64
 4   height     1038 non-null   float64
 5   weight     1038 non-null   float64
 6   smoke      1038 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 64.9+ KB


7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [174]:
print(baq["parity"].value_counts())

print("Pregnant women with no parity had a highest frequency")

No     755
Yes    283
Name: parity, dtype: int64
Pregnant women with no parity had a highest frequency


In [175]:
print(baq["smoke"].value_counts())
print("Pregnant women who were non smoker had a highest frequency")

Non-smoker    633
Smoker        405
Name: smoke, dtype: int64
Pregnant women who were non smoker had a highest frequency


In [176]:
print(pd.crosstab(baq["parity"], baq["smoke"]))
print("110 mothers were smoker who were pregnant before")

smoke   Non-smoker  Smoker
parity                    
No             460     295
Yes            173     110
110 mothers were smoker who were pregnant before


In [177]:
baq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1038 entries, 0 to 1085
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1038 non-null   int64  
 1   gestation  1038 non-null   float64
 2   parity     1038 non-null   object 
 3   age        1038 non-null   float64
 4   height     1038 non-null   float64
 5   weight     1038 non-null   float64
 6   smoke      1038 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 64.9+ KB


8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [178]:
baq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1038 entries, 0 to 1085
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1038 non-null   int64  
 1   gestation  1038 non-null   float64
 2   parity     1038 non-null   object 
 3   age        1038 non-null   float64
 4   height     1038 non-null   float64
 5   weight     1038 non-null   float64
 6   smoke      1038 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 64.9+ KB


In [179]:
print(f"The mean body weight of babies is", baq["bwt"].mean())

print(f"The median body weight of babies is", baq["bwt"].median()) # when organized the mid point of the series body weight is 120

print(f"The mode body weight of babies is", baq["bwt"].mode()) # most frequently observed value of the body weight is 117

The mean body weight of babies is 120.19942196531792
The median body weight of babies is 120.0
The mode body weight of babies is 0    117
dtype: int64


In [180]:
print(f"The mean gestation weeks of the mothers is", baq["gestation"].mean())

print(f"The median gestation weeks of the mothers is", baq["gestation"].median())
# when organized the mid point of the data series of gestation weeks is 280

print(f"The mode gestation weeks of the mothers is", baq["gestation"].mode())
# most frequently observed value of the gestation weeks is 282

The mean gestation weeks of the mothers is 279.9980732177264
The median gestation weeks of the mothers is 280.0
The mode gestation weeks of the mothers is 0    282.0
dtype: float64


In [181]:
print(f"The mean age of the mothers is", baq["age"].mean())

print(f"The median age of the mothers is", baq["age"].median())
# when organized the mid point of the data series of age is 26


print(f"The mode age of the mothers is", baq["age"].mode()) 
# most frequently observed value of the age is 24

The mean age of the mothers is 27.192678227360307
The median age of the mothers is 26.0
The mode age of the mothers is 0    24.0
dtype: float64


In [90]:
print(f"The mean height of the mothers is", baq["height"].mean())

print(f"The median height of the mothers is", baq["height"].median())
# when organized the mid point of the data series of height is 64

print(f"The mode height of the mothers is", baq["height"].mode())
# most frequently observed value of the height is 65

The mean height of the mothers is 64.00867052023122
The median height of the mothers is 64.0
The mode height of the mothers is 0    65.0
dtype: float64


In [182]:
print(f"The mean weight of the mothers is", baq["weight"].mean())

print(f"The median weight of the mothers is", baq["weight"].median())
# when organized the mid point of the data series of weight is 125

print(f"The mode weight of the mothers is", baq["weight"].mode())
# most frequently observed value of the weight is 130

The mean weight of the mothers is 125.67726396917148
The median weight of the mothers is 125.0
The mode weight of the mothers is 0    130.0
dtype: float64


9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [183]:
# range, standard deviation and percentile of birth weight
Range = baq["bwt"].max() - baq["bwt"].min()
stand_dev = baq["bwt"].std()
percentile = baq["bwt"].quantile(.85)

print(f"Range : {Range: .2f}")
print(f"Standard Deviation: {stand_dev:.2f}")
print(f"85th Percentile: {percentile:.2f}")

Range :  90.00
Standard Deviation: 16.40
85th Percentile: 137.00


In [184]:
# range, standard deviation and percentile of gestation
Range = baq["gestation"].max() - baq["gestation"].min()
stand_dev = baq["gestation"].std()
percentile = baq["gestation"].quantile(.85)

print(f"Range : {Range: .2f}")
print(f"Standard Deviation: {stand_dev:.2f}")
print(f"85th Percentile: {percentile:.2f}")

Range :  64.00
Standard Deviation: 11.39
85th Percentile: 292.00


In [185]:
# range, standard deviation and percentile of age
Range = baq["age"].max() - baq["age"].min()
stand_dev = baq["age"].std()
percentile = baq["age"].quantile(.85)

print(f"Range : {Range: .2f}")
print(f"Standard Deviation: {stand_dev:.2f}")
print(f"85th Percentile: {percentile:.2f}")

Range :  30.00
Standard Deviation: 5.77
85th Percentile: 34.00


In [186]:
# range, standard deviation and percentile of height
Range = baq["height"].max() - baq["height"].min()
stand_dev = baq["height"].std()
percentile = baq["height"].quantile(.85)

print(f"Range : {Range: .2f}")
print(f"Standard Deviation: {stand_dev:.2f}")
print(f"85th Percentile: {percentile:.2f}")

Range :  16.00
Standard Deviation: 2.48
85th Percentile: 67.00


In [187]:
# range, standard deviation and percentile of weight
Range = baq["weight"].max() - baq["weight"].min()
stand_dev = baq["weight"].std()
percentile = baq["weight"].quantile(.85)

print(f"Range : {Range: .2f}")
print(f"Standard Deviation: {stand_dev:.2f}")
print(f"85th Percentile: {percentile:.2f}")

Range :  90.00
Standard Deviation: 16.58
85th Percentile: 145.00


In [ ]:
# the birth weight of the newborne and height of the mother has the largest range
#the difference between the lowest and highest value of birth weight of the newborne is 90
#the difference between the lowest and highest value of height of the mother is 90

# age of mothers has a smallest range
#the difference between the lowest and highest value of age of the mother is 30

In [ ]:
# the weight of the mother has the highest standard deviation of 16.58 
# i.e. the measurement of the average distance between each data value and mean is 16.58


# the height of the mother has the smallest standard deviation of 2.48
# i.e. the measurement of the average distance between each data value and mean is 2.48
